src/Native/System.Native/pal_io.cpp

```diff
@@ -898,6 +903,132 @@ extern "C" int32_t SystemNative_Write(intptr_t fd, const void* buffer, int32_t b
     return static_cast<int32_t>(count);
 }
 
+#if !HAVE_FCOPYFILE
+// Read all data from inFd and write it to outFd
+static int32_t CopyFile_ReadWrite(int inFd, int outFd)
+{
+    // Allocate a buffer
+    const int BufferLength = 80 * 1024 * sizeof(char);
+    char* buffer = reinterpret_cast<char*>(malloc(BufferLength));
+    if (buffer == nullptr)
+    {
+        return -1;
+    }
+
+    // Repeatedly read from the source and write to the destination
+    while (true)
+    {
+        // Read up to what will fit in our buffer.  We're done if we get back 0 bytes.
+        ssize_t bytesRead;
+        while (CheckInterrupted(bytesRead = read(inFd, buffer, BufferLength)));
+        if (bytesRead == -1)
+        {
+            int tmp = errno;
+            free(buffer);
+            errno = tmp;
+            return -1;
+        }
+        if (bytesRead == 0)
+        {
+            break;
+        }
+        assert(bytesRead > 0);
+
+        // Write what was read.
+        ssize_t offset = 0;
+        while (bytesRead > 0)
+        {
+            ssize_t bytesWritten;
+            while (CheckInterrupted(bytesWritten = write(outFd, buffer + offset, static_cast<size_t>(bytesRead))));
+            if (bytesWritten == -1)
+            {
+                int tmp = errno;
+                free(buffer);
+                errno = tmp;
+                return -1;
+            }
+            assert(bytesWritten >= 0);
+            bytesRead -= bytesWritten;
+            offset += bytesWritten;
+        }
+    }
+
+    free(buffer);
+    return 0;
+}
+#endif // !HAVE_FCOPYFILE
+
+extern "C" int32_t SystemNative_CopyFile(intptr_t sourceFd, intptr_t destinationFd)
+{
+    int inFd = ToFileDescriptor(sourceFd);
+    int outFd = ToFileDescriptor(destinationFd);
+
+#if HAVE_FCOPYFILE
+    // If fcopyfile is available (OS X), try to use it, as the whole copy 
+    // can be performed in the kernel, without lots of unnecessary copying.
+    // Copy data and metadata.
+    return fcopyfile(inFd, outFd, nullptr, COPYFILE_ALL) == 0 ? 0 : -1;
+#else
+    // Get the stats on the source file.
+    int ret;
+    struct stat_ sourceStat;
+    bool copied = false;
+#if HAVE_SENDFILE
+    while (CheckInterrupted(ret = fstat_(inFd, &sourceStat)));
+    if (ret != 0)
+    {
+        return -1;
+    }
+
+    // If sendfile is available (Linux), try to use it, as the whole copy 
+    // can be performed in the kernel, without lots of unnecessary copying.
+    if (sendfile(outFd, inFd, 0, UnsignedCast(sourceStat.st_size)) >= 0)
```

Comment:
Found out that when compiling the native code on arm, it actually default to x64 instead of arm and will use the 64-bit version of sendfile. I'm fixing the build.sh script to do the right thing and this should go away.